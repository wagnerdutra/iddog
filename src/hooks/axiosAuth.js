import { useUser } from 'Root/context/user-context';
import axios from '../services/api';

function useAxiosAuth() {
  // const user = useUser();

  const axiosAuth = {
    get: url =>
      axios.get(url, {
        headers: {
          Authorization:
            'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJpZGRvZy1zZXJ2aWNlIiwic3ViIjoiNWQyZTUxZDc0NzRkYWFmZTAwYjIwMDM5IiwiaWF0IjoxNTYzMzE2Njk1LCJleHAiOjE1NjQ2MTI2OTV9.K_xepGlqlzIkzuZu1I5ScK3NspA7zAGmq9Y0zsOFe2w'
        }
      })
  };

  return axiosAuth;
}

export { useAxiosAuth };
