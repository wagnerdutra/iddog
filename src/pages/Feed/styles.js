import styled from 'styled-components';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  background: var(--main-bg-color);
  width: 100%;
  height: 100%;
`;

const Nav = styled.ul`
  margin-top: 20px;
  list-style: none;
  display: flex;

  li {
    margin-right: 10px;

    &.active {
      font-weight: bold;
    }

    a {
      color: inherit;
      text-decoration: none;
      font-size: 18px;
      line-height: 32px;
    }
  }
`;

const Content = styled.div`
  margin-top: 100px;
  justify-content: center;
  width: 80%;
  display: flex;
  align-items: center;
  justify-content: center;
  background: inherit;
  width: 100%;
`;

const ImagesList = styled.div`
  display: flex;
  flex-flow: row wrap;
  align-items: center;
  justify-content: center;
`;

export { Container, ImagesList, Content, Nav };
