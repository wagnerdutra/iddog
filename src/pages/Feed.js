import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import propTypes from 'prop-types';

import { categories } from 'Root/routes';

import { useAxiosAuth } from 'Root/hooks/axiosAuth';

import CardImage from 'Components/CardImage';
import Loading from 'Components/Loading';

import { Container, ImagesList, Content, Nav } from './styles';

const Header = React.memo(({ currentCategory }) => (
  <>
    <h1>THE IDDOG</h1>
    <Nav>
      {categories.map(category => (
        <li
          key={category.query}
          className={`${category.query === currentCategory ? 'active' : ''}`}
        >
          <Link to={`/feed?category=${category.query}`}>{category.name}</Link>
        </li>
      ))}
    </Nav>
  </>
));

Header.propTypes = {
  currentCategory: propTypes.string
};

Header.defaultProps = {
  currentCategory: categories.find(category => category.default).quert
};

function Feed({ location }) {
  const axiosAuth = useAxiosAuth();

  const [categoryImagesList, setCategoryImagesList] = useState({ category: '', list: [] });
  const [isLoaging, setIsLoading] = useState(false);
  const [currentCategory, setCurrentCategory] = useState(null);

  useEffect(() => {
    async function fetchCaregoryImageList() {
      const params = new URLSearchParams(location.search);
      const categoryQuery = params.get('category') || null;

      setIsLoading(true);

      const categoryRoute = categoryQuery || categories.find(category => category.default).query;

      setCurrentCategory(categoryRoute);

      const { data } = await axiosAuth.get(`/feed?category=${categoryRoute}`);

      setIsLoading(false);

      setCategoryImagesList(data);
    }

    fetchCaregoryImageList();
  }, [location.search]);

  return (
    <Container>
      <Header currentCategory={currentCategory} />
      <Content>
        {isLoaging ? (
          <Loading />
        ) : (
          <ImagesList>
            {categoryImagesList.list.length > 0 &&
              categoryImagesList.list.map(categoryImage => (
                <CardImage
                  src={categoryImage}
                  key={categoryImage}
                  width={300}
                  height={300}
                  margin={10}
                  cursor
                />
              ))}
          </ImagesList>
        )}
      </Content>
    </Container>
  );
}

Feed.propTypes = {
  location: propTypes.shape({
    search: propTypes.string.isRequired
  }).isRequired
};

export default Feed;
