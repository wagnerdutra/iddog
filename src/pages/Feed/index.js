import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import propTypes from 'prop-types';

import { categories } from 'Root/routes';

import { useAxiosAuth } from 'Root/hooks/axiosAuth';

import CardImage from 'Components/CardImage';
import Loading from 'Components/Loading';
import Modal from 'Components/Modal';
import CSSTransitionEF from 'Components/CSSTransitionEF';

import { Container, ImagesList, Content, Nav } from './styles';

const Header = React.memo(({ currentCategory }) => (
  <>
    <Nav>
      {categories.map(category => (
        <li
          key={category.query}
          className={`${category.query === currentCategory ? 'active' : ''}`}
        >
          <Link to={`/feed?category=${category.query}`}>{category.name}</Link>
        </li>
      ))}
    </Nav>
  </>
));

Header.propTypes = {
  currentCategory: propTypes.string
};

Header.defaultProps = {
  currentCategory: categories.find(category => category.default).quert
};

function Feed({ location }) {
  const axiosAuth = useAxiosAuth();

  const [categoryImagesList, setCategoryImagesList] = useState([]);
  const [isLoaging, setIsLoading] = useState(false);
  const [currentCategory, setCurrentCategory] = useState(null);

  useEffect(() => {
    async function fetchCaregoryImageList() {
      const params = new URLSearchParams(location.search);
      const categoryQuery = params.get('category') || null;

      const categoryRoute = categoryQuery || categories.find(category => category.default).query;

      setIsLoading(true);
      setCurrentCategory(categoryRoute);

      const {
        data: { list }
      } = await axiosAuth.get(`/feed?category=${categoryRoute}`);

      setIsLoading(false);
      setCategoryImagesList(list);
    }

    fetchCaregoryImageList();
  }, [location.search, axiosAuth]);

  const [modalProperties, setModal] = useState({
    showModal: false,
    imageSrc: ''
  });

  function closeModal() {
    setModal({ ...modalProperties, showModal: false });
  }

  function handleImageClick(img) {
    const { src: imageSrc } = img;
    setModal({
      showModal: true,
      imageSrc
    });
  }

  return (
    <Container>
      <Header currentCategory={currentCategory} />
      <Content>
        {isLoaging ? (
          <Loading />
        ) : (
          <ImagesList>
            {categoryImagesList.length > 0 &&
              categoryImagesList.map(categoryImage => (
                <CardImage
                  onClick={handleImageClick}
                  src={categoryImage}
                  key={categoryImage}
                  width={300}
                  height={300}
                  margin={10}
                  cursor
                />
              ))}
          </ImagesList>
        )}
      </Content>
      <Modal showModal={modalProperties.showModal} closeModal={closeModal}>
        <CardImage
          src={modalProperties.imageSrc}
          width={500}
          height={500}
          margin={10}
          hover={false}
        />
      </Modal>
    </Container>
  );
}

Feed.propTypes = {
  location: propTypes.shape({
    search: propTypes.string.isRequired
  }).isRequired
};

export default Feed;
