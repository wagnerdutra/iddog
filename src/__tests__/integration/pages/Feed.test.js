import React from 'react';

import { Route, Router } from 'react-router-dom';

import { createMemoryHistory } from 'history';

import { render, wait, fireEvent } from '@testing-library/react';

import Feed from 'Pages/Feed';

import { useAxiosAuth } from 'Root/hooks/axiosAuth';

jest.mock('Root/hooks/axiosAuth');

useAxiosAuth.mockReturnValue({
  get: url => {
    switch (url) {
      case '/feed?category=husky':
        return Promise.resolve({ data: { list: ['1'] } });
      case '/feed?category=labrador':
        return Promise.resolve({ data: { list: ['1', '2'] } });
      case '/feed?category=pug':
        return Promise.resolve({ data: { list: ['1', '2', '3'] } });
      case '/feed?category=hound':
        return Promise.resolve({ data: { list: ['1', '2', '3', '4'] } });
      default:
        return Promise.resolve({ data: { list: ['1'] } });
    }
  }
});

describe('<Feed />', () => {
  it('should handle the page Feed navigation', async () => {
    const route = '/feed';
    const history = createMemoryHistory({ initialEntries: [route] });
    const { container, getByText, debug } = render(
      <Router history={history}>
        <Route exact path="/feed" component={Feed} />
      </Router>
    );

    const huskyLink = getByText('Husky');
    const labradorLink = getByText('Labrador');
    const houndLink = getByText('Hound');
    const pugLink = getByText('Pug');

    await wait(() => {
      const imgs = container.querySelectorAll('img');
      expect(imgs.length).toBe(1);
    });

    fireEvent.click(pugLink);

    await wait(() => {
      const imgs = container.querySelectorAll('img');
      expect(imgs.length).toBe(3);
    });

    fireEvent.click(houndLink);

    await wait(() => {
      const imgs = container.querySelectorAll('img');
      expect(imgs.length).toBe(4);
    });

    fireEvent.click(labradorLink);

    await wait(() => {
      const imgs = container.querySelectorAll('img');
      expect(imgs.length).toBe(2);
    });

    fireEvent.click(huskyLink);

    await wait(() => {
      const imgs = container.querySelectorAll('img');
      expect(imgs.length).toBe(1);
    });
  });
});
