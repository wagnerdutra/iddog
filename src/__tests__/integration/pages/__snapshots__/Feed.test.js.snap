// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Feed /> should handle the page Feed navigation 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: var(--main-bg-color);
  width: 100%;
  height: 100%;
}

.c1 {
  margin-top: 20px;
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 li {
  margin-right: 10px;
}

.c1 li.active {
  font-weight: bold;
}

.c1 li a {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 18px;
  line-height: 32px;
}

.c2 {
  margin-top: 100px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 80%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: inherit;
  width: 100%;
}

<div>
  <div
    class="c0"
  >
    <ul
      class="c1"
    >
      <li
        class="active"
      >
        <a
          href="/feed?category=husky"
        >
          Husky
        </a>
      </li>
      <li
        class=""
      >
        <a
          href="/feed?category=labrador"
        >
          Labrador
        </a>
      </li>
      <li
        class=""
      >
        <a
          href="/feed?category=hound"
        >
          Hound
        </a>
      </li>
      <li
        class=""
      >
        <a
          href="/feed?category=pug"
        >
          Pug
        </a>
      </li>
    </ul>
    <div
      class="c2"
    >
      <img
        alt="loading"
        class=""
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;
